File: TOMIMSV4/DAL/Mantenimientos/Producto/ProductoParametros/clsLnProducto_parametros.vb
Imports System.Data.SqlClient

Public Class clsLnProducto_parametros

    Public Shared Sub Cargar(ByRef oBeProducto_parametros As clsBeProducto_parametros, ByRef dr As DataRow)

        Try

            With oBeProducto_parametros
                .IdProductoParametro = IIf(IsDBNull(dr.Item("IdProductoParametro")), 0, dr.Item("IdProductoParametro"))
                .IdParametro = IIf(IsDBNull(dr.Item("IdParametro")), 0, dr.Item("IdParametro"))
                .IdProducto = IIf(IsDBNull(dr.Item("IdProducto")), 0, dr.Item("IdProducto"))
                .Valor_texto = IIf(IsDBNull(dr.Item("valor_texto")), "", dr.Item("valor_texto"))
                .Valor_numerico = IIf(IsDBNull(dr.Item("valor_numerico")), 0.0, dr.Item("valor_numerico"))
                .Valor_fecha = IIf(IsDBNull(dr.Item("valor_fecha")), Date.Now, dr.Item("valor_fecha"))
                .Valor_logico = IIf(IsDBNull(dr.Item("valor_logico")), False, dr.Item("valor_logico"))
                .Capturar_siempre = IIf(IsDBNull(dr.Item("capturar_siempre")), False, dr.Item("capturar_siempre"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
            End With

        Catch ex As Exception
            Throw New Exception("Producto_parametros_Cargar: " & ex.message)
        End Try

    End Sub

    Public Shared Function Insertar(ByRef oBeProducto_parametros As clsBeProducto_parametros, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand

        Try

            Ins.Init("producto_parametros")
            Ins.Add("idproductoparametro", "@idproductoparametro", DataType.Parametro)
            Ins.Add("idparametro", "@idparametro", DataType.Parametro)
            Ins.Add("idproducto", "@idproducto", DataType.Parametro)
            Ins.Add("valor_texto", "@valor_texto", DataType.Parametro)
            Ins.Add("valor_numerico", "@valor_numerico", DataType.Parametro)
            Ins.Add("valor_fecha", "@valor_fecha", DataType.Parametro)
            Ins.Add("valor_logico", "@valor_logico", DataType.Parametro)
            Ins.Add("capturar_siempre", "@capturar_siempre", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)

            Dim sp As String = Ins.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeProducto_parametros.IdProductoParametro))
            cmd.Parameters.Add(New SqlParameter("@IDPARAMETRO", oBeProducto_parametros.IdParametro))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTO", oBeProducto_parametros.IdProducto))

            cmd.Parameters.Add(New SqlParameter("@VALOR_TEXTO", IIf(oBeProducto_parametros.Valor_texto.Trim = "", DBNull.Value, oBeProducto_parametros.Valor_texto)))
            cmd.Parameters.Add(New SqlParameter("@VALOR_NUMERICO", IIf(oBeProducto_parametros.Valor_numerico = 0, DBNull.Value, oBeProducto_parametros.Valor_numerico)))
            cmd.Parameters.Add(New SqlParameter("@VALOR_FECHA", IIf(oBeProducto_parametros.Valor_fecha = New Date(1900, 1, 1), DBNull.Value, oBeProducto_parametros.Valor_fecha)))
            cmd.Parameters.Add(New SqlParameter("@VALOR_LOGICO", IIf(oBeProducto_parametros.Valor_logico, True, oBeProducto_parametros.Valor_logico)))

            cmd.Parameters.Add(New SqlParameter("@CAPTURAR_SIEMPRE", oBeProducto_parametros.Capturar_siempre))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_parametros.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_parametros.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_parametros.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_parametros.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_parametros.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()
            Return rowsAffected

            oBeProducto_parametros.IdProductoParametro = CInt(cmd.Parameters("@IDPRODUCTOPARAMETRO").Value)

        Catch ex As Exception
            Throw New Exception("Producto_parametros_Insertar: " & ex.Message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeProducto_parametros As clsBeProducto_parametros, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand

        Try

            Upd.Init("producto_parametros")
            Upd.Add("idproductoparametro", "@idproductoparametro", DataType.Parametro)
            Upd.Add("idparametro", "@idparametro", DataType.Parametro)
            Upd.Add("idproducto", "@idproducto", DataType.Parametro)
            Upd.Add("valor_texto", "@valor_texto", DataType.Parametro)
            Upd.Add("valor_numerico", "@valor_numerico", DataType.Parametro)
            Upd.Add("valor_fecha", "@valor_fecha", DataType.Parametro)
            Upd.Add("valor_logico", "@valor_logico", DataType.Parametro)
            Upd.Add("capturar_siempre", "@capturar_siempre", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Where("IdProductoParametro = @IdProductoParametro")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeProducto_parametros.IdProductoParametro))
            cmd.Parameters.Add(New SqlParameter("@IDPARAMETRO", oBeProducto_parametros.IdParametro))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTO", oBeProducto_parametros.IdProducto))

            cmd.Parameters.Add(New SqlParameter("@VALOR_TEXTO", IIf(oBeProducto_parametros.Valor_texto.Trim = "", DBNull.Value, oBeProducto_parametros.Valor_texto)))
            cmd.Parameters.Add(New SqlParameter("@VALOR_NUMERICO", IIf(oBeProducto_parametros.Valor_numerico = 0, DBNull.Value, oBeProducto_parametros.Valor_numerico)))
            cmd.Parameters.Add(New SqlParameter("@VALOR_FECHA", IIf(oBeProducto_parametros.Valor_fecha = New Date(1900, 1, 1), DBNull.Value, oBeProducto_parametros.Valor_fecha)))
            cmd.Parameters.Add(New SqlParameter("@VALOR_LOGICO", IIf(oBeProducto_parametros.Valor_logico, True, oBeProducto_parametros.Valor_logico)))

            cmd.Parameters.Add(New SqlParameter("@CAPTURAR_SIEMPRE", oBeProducto_parametros.Capturar_siempre))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_parametros.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_parametros.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_parametros.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_parametros.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_parametros.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()
            Return rowsAffected

        Catch ex As Exception
            Throw New Exception("Producto_parametros_Actualizar: " & ex.Message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Function Eliminar(ByRef oBeProducto_parametros As clsBeProducto_parametros, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            cmd.CommandType = CommandType.Text


            Dim sp As String = " Delete from Producto_parametros" &
             "  Where(IdProductoParametro = @IdProductoParametro)"


            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then

                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else

                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)


            End If


            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeProducto_parametros.IdProductoParametro))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()
            Return rowsAffected

        Catch ex As Exception
            Throw New Exception("Producto_parametros_Eliminar: " & ex.message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
            cmd.Dispose()
        End Try
    End Function

    Public Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Producto_parametros"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex As Exception
            Throw New Exception("Producto_parametros_Listar: " & ex.message)
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBeProducto_parametros As clsBeProducto_parametros) As Boolean

        Try

            Dim sp As String = " SELECT * FROM Producto_parametros" &
                               " Where(IdProductoParametro = @IdProductoParametro)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim dad As New SqlDataAdapter(cmd)


            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeProducto_parametros.IdProductoParametro))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeProducto_parametros, dt.Rows(0))
            Else
                Throw New Exception("No se pudo obtener el registro")
            End If

            Return True

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBeProducto_parametros As clsBeProducto_parametros,
                                   ByVal lConnection As SqlConnection,
                                   ByVal lTransaction As SqlTransaction) As Boolean

        Obtener = False

        Try

            Dim sp As String = " SELECT * FROM Producto_parametros" &
                               " Where(IdProductoParametro = @IdProductoParametro)"

            Dim cmd As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)

            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETRO", oBeProducto_parametros.IdProductoParametro))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeProducto_parametros, dt.Rows(0))
                Return True
            End If

        Catch ex As Exception
            Throw ex
        End Try

    End Function

End Class

File: WMSWebAPI/Dal/Producto/clsLnProducto_parametros.cs
using Microsoft.Data.SqlClient;
using System.Data;
using System.Diagnostics;
using WMSWebAPI.Entity.Producto;

public class clsLnProductoParametros
{
    private static readonly clsInsert Inserter = new clsInsert();
    private static readonly clsUpdate Updater = new clsUpdate();

    private static string GetConnectionString(IConfiguration config) =>
        config.GetConnectionString("CST") ?? throw new InvalidOperationException("Cadena de conexión 'CST' no está configurada.");
    private static SqlCommand CreateCommand(string sql, SqlConnection conn, SqlTransaction? tx = null) =>
        new SqlCommand(sql, conn, tx) { CommandType = CommandType.Text };
    private static void BindParams(SqlCommand cmd, clsBeProducto_parametros e)
    {
        cmd.Parameters.AddWithValue("@IdProductoParametro", e.IdProductoParametro);
        cmd.Parameters.AddWithValue("@IdParametro", e.IdParametro);
        cmd.Parameters.AddWithValue("@IdProducto", e.IdProducto);
        cmd.Parameters.AddWithValue("@valor_texto", e.Valor_texto);
        cmd.Parameters.AddWithValue("@valor_numerico", e.Valor_numerico);
        cmd.Parameters.AddWithValue("@valor_fecha", e.Valor_fecha);
        cmd.Parameters.AddWithValue("@valor_logico", e.Valor_logico);
        cmd.Parameters.AddWithValue("@capturar_siempre", e.Capturar_siempre);
        cmd.Parameters.AddWithValue("@user_agr", e.User_agr);
        cmd.Parameters.AddWithValue("@fec_agr", e.Fec_agr);
        cmd.Parameters.AddWithValue("@user_mod", e.User_mod);
        cmd.Parameters.AddWithValue("@fec_mod", e.Fec_mod);
        cmd.Parameters.AddWithValue("@activo", e.Activo);
    }
    private static void LoadFromRow(ref clsBeProducto_parametros e, DataRow r)
    {
        e.IdProductoParametro = r.Field<int>("IdProductoParametro");
        e.IdParametro = r.Field<int>("IdParametro");
        e.IdProducto = r.Field<int>("IdProducto");
        e.Valor_texto = r.Field<string>("valor_texto") ?? "";
        e.Valor_numerico = r.Field<double?>("valor_numerico") ?? 0;
        e.Valor_fecha = r.Field<DateTime?>("valor_fecha") ?? DateTime.Now;
        e.Valor_logico = r.Field<bool?>("valor_logico") ?? false;
        e.Capturar_siempre = r.Field<bool?>("capturar_siempre") ?? false;
        e.User_agr = r.Field<string>("user_agr") ?? "";
        e.Fec_agr = r.Field<DateTime?>("fec_agr") ?? DateTime.Now;
        e.User_mod = r.Field<string>("user_mod") ?? "";
        e.Fec_mod = r.Field<DateTime?>("fec_mod") ?? DateTime.Now;
        e.Activo = r.Field<bool?>("activo") ?? false;
    }
    private static void LoadFromReader(ref clsBeProducto_parametros e, IDataRecord r)
    {
        int GetInt(string col) => r.IsDBNull(r.GetOrdinal(col)) ? 0 : r.GetInt32(r.GetOrdinal(col));
        double GetDouble(string col) => r.IsDBNull(r.GetOrdinal(col)) ? 0 : r.GetDouble(r.GetOrdinal(col));
        string GetString(string col) => r.IsDBNull(r.GetOrdinal(col)) ? "" : r.GetString(r.GetOrdinal(col));
        DateTime GetDate(string col) => r.IsDBNull(r.GetOrdinal(col)) ? DateTime.Now : r.GetDateTime(r.GetOrdinal(col));
        bool GetBool(string col) => !r.IsDBNull(r.GetOrdinal(col)) && r.GetBoolean(r.GetOrdinal(col));

        e.IdProductoParametro = GetInt("IdProductoParametro");
        e.IdParametro = GetInt("IdParametro");
        e.IdProducto = GetInt("IdProducto");
        e.Valor_texto = GetString("valor_texto");
        e.Valor_numerico = GetDouble("valor_numerico");
        e.Valor_fecha = GetDate("valor_fecha");
        e.Valor_logico = GetBool("valor_logico");
        e.Capturar_siempre = GetBool("capturar_siempre");
        e.User_agr = GetString("user_agr");
        e.Fec_agr = GetDate("fec_agr");
        e.User_mod = GetString("user_mod");
        e.Fec_mod = GetDate("fec_mod");
        e.Activo = GetBool("activo");
    }
    public static int Insert(IConfiguration config, clsBeProducto_parametros e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        Inserter.Init("producto_parametros");
        Inserter.Add("idproductoparametro", "@IdProductoParametro", "F");
        Inserter.Add("idparametro", "@IdParametro", "F");
        Inserter.Add("idproducto", "@IdProducto", "F");
        Inserter.Add("valor_texto", "@valor_texto", "F");
        Inserter.Add("valor_numerico", "@valor_numerico", "F");
        Inserter.Add("valor_fecha", "@valor_fecha", "F");
        Inserter.Add("valor_logico", "@valor_logico", "F");
        Inserter.Add("capturar_siempre", "@capturar_siempre", "F");
        Inserter.Add("user_agr", "@user_agr", "F");
        Inserter.Add("fec_agr", "@fec_agr", "F");
        Inserter.Add("user_mod", "@user_mod", "F");
        Inserter.Add("fec_mod", "@fec_mod", "F");
        Inserter.Add("activo", "@activo", "F");

        var sql = Inserter.SQL();
        bool isTx = conn != null && tx != null;

        using var connection = isTx ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;

        try
        {
            if (!isTx)
            {
                connection.Open();
                localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            var cmd = CreateCommand(sql, connection, isTx ? tx : localTx);
            BindParams(cmd, e);

            int result = cmd.ExecuteNonQuery();

            if (!isTx) localTx?.Commit();
            return result;
        }
        catch (Exception ex)
        {
            if (!isTx) localTx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!isTx && connection.State == ConnectionState.Open)
                connection.Close();
        }
    }
    public static int Update(IConfiguration config, clsBeProducto_parametros e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        Updater.Init("producto_parametros");
        Updater.Add("idparametro", "@IdParametro", "F");
        Updater.Add("idproducto", "@IdProducto", "F");
        Updater.Add("valor_texto", "@valor_texto", "F");
        Updater.Add("valor_numerico", "@valor_numerico", "F");
        Updater.Add("valor_fecha", "@valor_fecha", "F");
        Updater.Add("valor_logico", "@valor_logico", "F");
        Updater.Add("capturar_siempre", "@capturar_siempre", "F");
        Updater.Add("user_agr", "@user_agr", "F");
        Updater.Add("fec_agr", "@fec_agr", "F");
        Updater.Add("user_mod", "@user_mod", "F");
        Updater.Add("fec_mod", "@fec_mod", "F");
        Updater.Add("activo", "@activo", "F");
        Updater.Where("IdProductoParametro = @IdProductoParametro");

        var sql = Updater.SQL();
        bool isTx = conn != null && tx != null;

        using var connection = isTx ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;

        try
        {
            if (!isTx)
            {
                connection.Open();
                localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            var cmd = CreateCommand(sql, connection, isTx ? tx : localTx);
            BindParams(cmd, e);

            int result = cmd.ExecuteNonQuery();

            if (!isTx) localTx?.Commit();
            return result;
        }
        catch (Exception ex)
        {
            if (!isTx) localTx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!isTx && connection.State == ConnectionState.Open)
                connection.Close();
        }
    }
    public static bool Existe(int id, SqlConnection conn, SqlTransaction tx)
    {
        const string sql = "SELECT COUNT(1) FROM Producto_parametros WHERE IdProductoParametro = @IdProductoParametro";

        try
        {
            using var cmd = new SqlCommand(sql, conn, tx);
            cmd.Parameters.AddWithValue("@IdProductoParametro", id);
            return Convert.ToInt32(cmd.ExecuteScalar()) > 0;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int InsertOrUpdate(IConfiguration config, clsBeProducto_parametros e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        using var connection = conn ?? new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = tx ?? connection.BeginTransaction();
        if (conn == null) connection.Open();

        try
        {
            if (Existe(e.IdProductoParametro, connection, localTx))
                return Update(config, e, connection, localTx);
            else
                return Insert(config, e, connection, localTx);
        }
        catch
        {
            if (conn == null) localTx.Rollback();
            throw;
        }
        finally
        {
            if (conn == null) { localTx.Commit(); connection.Close(); }
        }
    }
    public static List<clsBeProducto_parametros> GetAll(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_parametros";
        var list = new List<clsBeProducto_parametros>();

        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();

        SqlTransaction? tx = null;
        try
        {
            tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = new SqlCommand(sql, connection, tx);
            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                var obj = new clsBeProducto_parametros();
                LoadFromReader(ref obj, reader);
                list.Add(obj);
            }

            tx.Commit();
            return list;
        }
        catch (Exception ex)
        {
            tx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static bool GetSingle(IConfiguration config, ref clsBeProducto_parametros entity)
    {
        const string sql = "SELECT * FROM Producto_parametros WHERE IdProductoParametro = @IdProductoParametro";

        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();

        SqlTransaction? tx = null;
        try
        {
            tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = new SqlCommand(sql, connection, tx);
            cmd.Parameters.AddWithValue("@IdProductoParametro", entity.IdProductoParametro);

            using var adapter = new SqlDataAdapter(cmd);
            var table = new DataTable();
            adapter.Fill(table);

            tx.Commit();

            if (table.Rows.Count == 1)
            {
                LoadFromRow(ref entity, table.Rows[0]);
                return true;
            }

            return false;
        }
        catch (Exception ex)
        {
            tx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int MaxId(IConfiguration config, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        const string sql = "SELECT ISNULL(MAX(IdProductoParametro), 0) FROM Producto_parametros";
        bool isExternal = conn != null && tx != null;

        using var connection = isExternal ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;

        try
        {
            if (!isExternal)
            {
                connection.Open();
                localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            using var cmd = new SqlCommand(sql, connection, isExternal ? tx! : localTx!);
            var result = cmd.ExecuteScalar();

            if (!isExternal) localTx?.Commit();

            return result != null ? Convert.ToInt32(result) : 0;
        }
        catch (Exception ex)
        {
            if (!isExternal) localTx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
}